src/System.Private.CoreLib/shared/System/IO/Stream.cs

```diff
@@ -189,6 +189,65 @@ private int GetCopyBufferSize()
             return bufferSize;
         }
 
+        public virtual void CopyTo(ReadOnlySpanAction<byte, object> callback, object state, int bufferSize) =>
+            CopyTo(new WriteCallbackStream(callback, state), bufferSize);
+
+        public virtual Task CopyToAsync(Func<object, ReadOnlyMemory<byte>, CancellationToken, Task> callback, object state, int bufferSize, CancellationToken cancellationToken) =>
+            CopyToAsync(new WriteCallbackStream(callback, state), bufferSize, cancellationToken);
+
+        private sealed class WriteCallbackStream : Stream
+        {
+            private readonly ReadOnlySpanAction<byte, object>? _action;
+            private readonly Func<object, ReadOnlyMemory<byte>, CancellationToken, Task>? _func;
+            private readonly object _state;
+
+            public WriteCallbackStream(ReadOnlySpanAction<byte, object> action, object state)
+            {
+                _action = action ?? throw new ArgumentNullException(nameof(action));
+                _state = state;
+            }
+
+            public WriteCallbackStream(Func<object, ReadOnlyMemory<byte>, CancellationToken, Task> func, object state)
+            {
+                _func = func ?? throw new ArgumentNullException(nameof(func));
+                _state = state;
+            }
+
+            public override void Write(byte[] buffer, int offset, int count)
+            {
+                if (_action == null) throw new NotSupportedException();
```

Comment:
Open question: Are we sure we want an exception in this code path? It's possible (though unlikely) that this method might be called if the outer `Stream` doesn't support async.

(Same comment applies to other `NotSupportException`-throwing methods in this code.)

Reply:
My assumption is that we expect `CopyTo` overrides to call only `Write` (and maybe `Flush`) and `CopyToAsync` ones to call only `WriteAsync` (and maybe `Flush` via `FlushAsync`).
Do we want to support "cross combinations" and implement sync-to-async and async-to-sync wrappers?
```suggestion
                if (_action != null)
                    _action(new ReadOnlySpan<byte>(buffer, offset, count), _state);
                else if (_func != null)
                    _func(new ReadOnlyMemory<byte>(buffer, offset, count), _state, CancellationToken.None).GetAwaiter().GetResult();
                else
                    throw new NotSupportedException();
```

Reply:
I don't know, hence the open question for others to chime in. There's probably nothing wrong with shipping this as-is and reconsidering this scenario if we get future reports that this is a real-world problem.

Reply:
Since it's relatively easy to add support for these mismatched scenarios I would lean towards adding that support now rather than later. Primarily because the source stream type often isn't known/controlled by the `CopyToAsync` caller and if the source happened to have such mismatched calls they would work today, but could break unexpectedly if changing to these new overloads.

Reply:
@stephentoub An opinion on this "open question"?

Reply:
I think we should make it "just work".

Reply:
See below for my "just work" commit. Is it enough ?